pkg/backend/filestate/bucket.go

```diff
@@ -4,14 +4,58 @@ import (
 	"context"
 	"io"
 	"path"
+	"path/filepath"
 
 	"github.com/pkg/errors"
 	"github.com/pulumi/pulumi/pkg/util/logging"
 	"gocloud.dev/blob"
 )
 
+// Bucket is a wrapper around an underlying gocloud blob.Bucket.  It ensures that we pass all paths
+// to it normalized to forward-slash form like it requires.
+type Bucket interface {
+	Copy(ctx context.Context, dstKey, srcKey string, opts *blob.CopyOptions) (err error)
+	Delete(ctx context.Context, key string) (err error)
+	List(opts *blob.ListOptions) *blob.ListIterator
+	SignedURL(ctx context.Context, key string, opts *blob.SignedURLOptions) (string, error)
+	ReadAll(ctx context.Context, key string) (_ []byte, err error)
+	WriteAll(ctx context.Context, key string, p []byte, opts *blob.WriterOptions) (err error)
+}
+
+// wrappedBucket encapsulates a true gocloud blob.Bucket, but ensures that all paths we send to it
+// are appropriately normalized to use forward slashes as required by it.  Without this, we may use
+// filepath.join which can make paths like `c:\temp\etc`.  gocloud's fileblob then converts those
+// backslashes to the hex string __0x5c__, breaking things on windows completely.
+type wrappedBucket struct {
+	bucket *blob.Bucket
+}
+
+func (b *wrappedBucket) Copy(ctx context.Context, dstKey, srcKey string, opts *blob.CopyOptions) (err error) {
+	return b.bucket.Copy(ctx, filepath.ToSlash(dstKey), filepath.ToSlash(srcKey), opts)
+}
+
+func (b *wrappedBucket) Delete(ctx context.Context, key string) (err error) {
+	return b.bucket.Delete(ctx, filepath.ToSlash(key))
+}
+
+func (b *wrappedBucket) List(opts *blob.ListOptions) *blob.ListIterator {
+	return b.bucket.List(opts)
+}
+
+func (b *wrappedBucket) SignedURL(ctx context.Context, key string, opts *blob.SignedURLOptions) (string, error) {
+	return b.bucket.SignedURL(ctx, filepath.ToSlash(key), opts)
+}
+
+func (b *wrappedBucket) ReadAll(ctx context.Context, key string) (_ []byte, err error) {
+	return b.bucket.ReadAll(ctx, filepath.ToSlash(key))
+}
+
+func (b *wrappedBucket) WriteAll(ctx context.Context, key string, p []byte, opts *blob.WriterOptions) (err error) {
+	return b.bucket.WriteAll(ctx, filepath.ToSlash(key), p, opts)
+}
```

Comment:
@ellismg let me know if you think this is an appropriate way to do things in 'go'.  Main problem is just that all the filepath APIs we use generate backslash windows paths.  But the gocloud blob apis take all those backslashes and make __5c__ out of them.  gocloud wants forward slashes for the paths that it ends up writing to the fs, regardless of platform.

this isolates us from the two behaviors and ensures no one accidentally passes the wrong form to gocloud.

Reply:
Seems like the best fix.